File: 0050-powx-n/0050-powx-n.java
class Solution {
    private double calculate(double x, long n){
        if(n == 0){
            return 1;
        }
        if(n == 1){
            return x;
        }
        double answer = 1;
        double half = calculate(x, n / 2);
        if(n % 2 == 1) answer = x;
        return answer * half * half;
    }
    public double myPow(double x, int n) {
        double answer = 1;
        long N = n;
        if(N < 0){
            x = 1/x;
            N = -N;
        }
        answer = calculate(x, N);
        return answer;
    }
}

File: 0050-powx-n/0050-powx-n.py
class Solution:
    def pow(self,x,n):
        if n == 0:
            return x
        carry = 1
        left = pow(x,n//2)
        right = pow(x,n//2)
        if n % 2 !=0:
            carry = x
        return left * right * ans

    def myPow(self, x: float, n: int) -> float:
        # temp = 0
        # infinity = [-2147483648,2147483648]
        # if n in infinity:
        #     if x in [0,1,-1]:
        #         return x
        #     return 0
        if n < 0:
            x = 1/x
            n = -n
        ans = pow(x,n)
        # if temp == 1:
        #     ans = 1/ans
        return ans